### PR TITLE
Added search component

### PR DESCRIPTION
**Pull request by Yashgupta9330**
made global search component https://github.com/OWASP/Nest/issues/55

### CODE REVIEW

**Comment by Yashgupta9330** [comment]
In the previous one by mistake i closed the pr

**Review by arkid15r — CHANGES_REQUESTED** [review]
Could you address all items listed in the old PR's comment -- https://github.com/OWASP/Nest/pull/115#pullrequestreview-2476966266


Screenshot 2024-12-06 at 10 31 32 AM
Screenshot 2024-12-06 at 10 32 00 AM

**Comment by Yashgupta9330** [comment]
Okk I did some changes i added the debouncing part rest part i will do it

File: frontend/src/pages/Projects.tsx [hunk 1]
```diff
@@ -23,8 +26,14 @@ export default function Projects() {
   }, [])
 
   return (
-    <div className=" w-full min-h-screen flex flex-col justify-normal items-center text-text p-5 md:p-20 ">
-      <div className=" w-full h-fit flex flex-col justify-normal items-center gap-4 ">
+    <div className="w-full min-h-screen flex flex-col justify-normal items-center text-text p-5 md:p-20">
+      <SearchBar
+        placeholder="Search for OWASP projects..."
+        searchEndpoint="http://localhost:8000/api/v1/owasp/search/projects"
```
Comment (harsh3dev) [inline frontend/src/pages/Projects.tsx]:
```suggestion
        searchEndpoint="http://localhost:8000/api/v1/owasp/search/project"
```

File: frontend/src/pages/Projects.tsx [hunk 2]
```diff
@@ -23,8 +26,14 @@ export default function Projects() {
   }, [])
 
   return (
-    <div className=" w-full min-h-screen flex flex-col justify-normal items-center text-text p-5 md:p-20 ">
-      <div className=" w-full h-fit flex flex-col justify-normal items-center gap-4 ">
+    <div className="w-full min-h-screen flex flex-col justify-normal items-center text-text p-5 md:p-20">
```
Comment (harsh3dev) [inline frontend/src/pages/Projects.tsx]:
```suggestion
    <div className="w-full min-h-screen flex flex-col justify-normal items-center text-text px-5 md:px-20">
```

File: frontend/src/components/Search.tsx [hunk 3]
```diff
@@ -0,0 +1,86 @@
...
+          setLoading(false)
+        }
+      }
+
+      fetchResults()
+    }
+  }, [debouncedQuery, searchEndpoint, onSearchResult])
+
+  return (
+    <div className="w-full max-w-md mx-auto mt-8">
```
Comment (harsh3dev) [inline frontend/src/components/Search.tsx]:
```suggestion
    <div className="w-full max-w-md mx-auto my-8">
```

**Comment by harsh3dev** [comment]
@Yashgupta9330 
* It will be better to use the `.env` variables for the API endpoints
* Also, when the search is empty, i.e., there is no value, the Project data should be reset to the default.

Also, @arkid15r, should we add dynamic URL functionality now? When the user comes to this URL, for example, `/projects?q=owasp,` then automatically, the relevant result is shown.
Shall this be implemented now or later?

**Comment by arkid15r** [comment]
> Also, @arkid15r, should we add dynamic URL functionality now? When the user comes to this URL, for example, `/projects?q=owasp,` then automatically, the relevant result is shown. Shall this be implemented now or later?

I expected it to be implemented as a part of this component (see my [comment](https://github.com/OWASP/Nest/pull/115#pullrequestreview-2476966266) in the previous PR). 
@harsh3dev thanks for helping w/ the review btw.

@Yashgupta9330 do you think it's something we can have soon (today/tomorrow) or you'd prefer this part to be extracted to a separate issue and done in a different PR?

**Comment by Yashgupta9330** [comment]
@arkid15r  can i do dynamic url search in a different pr?

**Comment by arkid15r** [comment]
> @arkid15r can i do dynamic url search in a different pr?

Sure. Just keep in mind it's a vital user facing feature and we 100% need it in the new frontend.

**Comment by Yashgupta9330** [comment]
@arkid15r  I'm thinking of not using the API for search. Since we already have the projects data loaded initially, we can filter the data locally instead of making additional API calls

**Comment by arkid15r** [comment]
> @arkid15r I'm thinking of not using the API for search. Since we already have the projects data loaded initially, we can filter the data locally instead of making additional API calls

No, that'd not be the right way to implement it. We need those results to be generated and ranked by our index engine. This can't be achieved by local filtering.

**Review by arkid15r — COMMENTED** [review]
Another discrepancy I've noticed: it doesn't populate the default project list back after I delete the search query (empty search box):
  - enter "nest", there should be 1 search result
  - remove "nest", there should be more than 1 search result (default view)

**Review by arkid15r — COMMENTED** [review]
The behavior looks good expect the search bar looses focus after rendering the search results. Could you look into that?

Also we now have frontend tests in CI/CD. Please merge the changes from main and make sure it's green.

Thank you @Yashgupta9330

File: frontend/src/components/Search.tsx [hunk 4]
```diff
@@ -0,0 +1,94 @@
...
+      return;
+    }
+
+    setLoading(true)
+    setError(null)
+    try {
+      const response = await axios.get(searchEndpoint, {
+        params: { q: searchQuery },
+      })
+      console.log(response.data)
```
Comment (arkid15r) [inline frontend/src/components/Search.tsx]: Please remove excessive prints

File: frontend/src/pages/Projects.tsx [hunk 5]
```diff
@@ -23,10 +26,16 @@ export default function Projects() {
   }, [])
 
   return (
-    <div className=" w-full min-h-screen flex flex-col justify-normal items-center text-text p-5 md:p-20 ">
-      <div className=" w-full h-fit flex flex-col justify-normal items-center gap-4 ">
+    <div className="w-full min-h-screen flex flex-col justify-normal items-center text-text p-5">
+      <div className="w-full h-fit flex flex-col justify-normal items-center gap-4">
+        <SearchBar
+          placeholder="Search for OWASP projects..."
+          searchEndpoint={`${import.meta.env.VITE_NEST_API_URL}/owasp/search/project`}
```
Comment (arkid15r) [inline frontend/src/pages/Projects.tsx]: Could you try import { API_URL } from '../utils/credentials.ts' instead?

**Comment by yashpandey06** [comment]
> The behavior looks good expect the search bar looses focus after rendering the search results. Could you look into that?
> 
> Also we now have frontend tests in CI/CD. Please merge the changes from main and make sure it's green.
> 
> Thank you @Yashgupta9330 

@arkid15r why we haven't tests for the new component ? Here...

**Comment by arkid15r** [comment]
> @arkid15r why we haven't added tests for the new component ? Here...

That's a fair question. I guess it's not done the same reason why [this PR](https://github.com/OWASP/Nest/pull/116) had been merged without implementing dockerized FE tests -- we want consistent progress of gradual improvements in order to migrate to the new FE. However, our ultimate goal is having proper test coverage and it will be achieved eventually. 

As you can see in the issue history we agreed to merge this PR as just a part of the search bar implementation. There will be following PRs and one of them will address your testing concerns.

**Comment by yashpandey06** [comment]
> > @arkid15r why we haven't added tests for the new component ? Here...
> 
> That's a fair question. I guess it's not done the same reason why [this PR](https://github.com/OWASP/Nest/pull/116) had been merged without implementing dockerized FE tests -- we want consistent progress of gradual improvements in order to migrate to the new FE. However, our ultimate goal is having proper test coverage and it will be achieved eventually. 
> 
> As you can see in the issue history we agreed to merge this PR as just a part of the search bar implementation. There will be following PRs and one of them will address your testing concerns.
> 
> 

Thanks for the clarification, appreciate that !

**Comment by arkid15r** [comment]
@Yashgupta9330 we configured some code quality checks. Please resolve the conflicts and use `make check` & `make test` to make sure it's green.

File: frontend/eslint.config.js [hunk 6]
```diff
@@ -50,7 +48,7 @@ export default [
       '@typescript-eslint/no-explicit-any': 'warn',
       '@typescript-eslint/no-inferrable-types': 'warn',
       '@typescript-eslint/no-unused-expressions': 'error',
-      '@typescript-eslint/no-unused-vars': ['warn', { argsIgnorePattern: '^_' }],
+      '@typescript-eslint/no-unused-vars': 'off',
```
Comment (arkid15r) [inline frontend/eslint.config.js]: There is no need to change this rule. Could you revert it?

**Review by arkid15r — CHANGES_REQUESTED** [review]
Hey @Yashgupta9330 
could you address all my comments including console logging removal and search bar focus.

It'd also be great not to suppress linting rules if possible (`// eslint-disable-next-line no-unused-vars`)

Thank you

**Comment by Yashgupta9330** [comment]
> Hey @Yashgupta9330 could you address all my comments including console logging removal and search bar focus.
> 
> It'd also be great not to suppress linting rules if possible (`// eslint-disable-next-line no-unused-vars`)
> 
> Thank you

Actually, I am using a function prop name, but it is not being used in my code, and that is why this error is occurring.

File: frontend/eslint.config.js (hunk 6 above)
Comment (Yashgupta9330) [reply frontend/eslint.config.js]: fixed this

File: frontend/src/components/Search.tsx (hunk 4 above)
Comment (Yashgupta9330) [reply frontend/src/components/Search.tsx]: removed

File: README.md [hunk 7]
```diff
@@ -7,6 +7,28 @@ Follow these steps to set up the OWASP Nest application:
 1. **Clone the Repository**:
    - Clone the repository code from GitHub using the following command:
 
+1. Run `make setup` to create a super user.
```
Comment (arkid15r) [inline README.md]: Why this change?

**Review by arkid15r — COMMENTED** [review]
I'm unable to run the frontend for this PR:

```
nest-frontend  | /home/owasp/node_modules/rollup/dist/native.js:63
nest-frontend  |                throw new Error(
nest-frontend  |                      ^
nest-frontend  | 
nest-frontend  | Error: Cannot find module @rollup/rollup-linux-arm64-gnu. npm has a bug related to optional dependencies (https://github.com/npm/cli/issues/4828). Please try `npm i` again after removing both package-lock.json and node_modules directory.
nest-frontend  |     at requireWithFriendlyError (/home/owasp/node_modules/rollup/dist/native.js:63:9)
nest-frontend  |     at Object.<anonymous> (/home/owasp/node_modules/rollup/dist/native.js:72:76)
nest-frontend  |     at Module._compile (node:internal/modules/cjs/loader:1565:14)
nest-frontend  |     at Object..js (node:internal/modules/cjs/loader:1708:10)
nest-frontend  |     at Module.load (node:internal/modules/cjs/loader:1318:32)
nest-frontend  |     at Function._load (node:internal/modules/cjs/loader:1128:12)
nest-frontend  |     at TracingChannel.traceSync (node:diagnostics_channel:322:14)
nest-frontend  |     at wrapModuleLoad (node:internal/modules/cjs/loader:219:24)
nest-frontend  |     at cjsLoader (node:internal/modules/esm/translators:263:5)
nest-frontend  |     at ModuleWrap.<anonymous> (node:internal/modules/esm/translators:196:7) {
nest-frontend  |   [cause]: Error: Cannot find module '@rollup/rollup-linux-arm64-gnu'
nest-frontend  |   Require stack:
nest-frontend  |   - /home/owasp/node_modules/rollup/dist/native.js
nest-frontend  |       at Function._resolveFilename (node:internal/modules/cjs/loader:1249:15)
nest-frontend  |       at Function._load (node:internal/modules/cjs/loader:1075:27)
nest-frontend  |       at TracingChannel.traceSync (node:diagnostics_channel:322:14)
nest-frontend  |       at wrapModuleLoad (node:internal/modules/cjs/loader:219:24)
nest-frontend  |       at Module.require (node:internal/modules/cjs/loader:1340:12)
nest-frontend  |       at require (node:internal/modules/helpers:138:16)
nest-frontend  |       at requireWithFriendlyError (/home/owasp/node_modules/rollup/dist/native.js:45:10)
nest-frontend  |       at Object.<anonymous> (/home/owasp/node_modules/rollup/dist/native.js:72:76)
nest-frontend  |       at Module._compile (node:internal/modules/cjs/loader:1565:14)
nest-frontend  |       at Object..js (node:internal/modules/cjs/loader:1708:10) {
nest-frontend  |     code: 'MODULE_NOT_FOUND',
nest-frontend  |     requireStack: [ '/home/owasp/node_modules/rollup/dist/native.js' ]
nest-frontend  |   }
nest-frontend  | }
nest-frontend  | 
nest-frontend  | Node.js v22.12.0
nest-frontend exited with code 1
```

**Comment by Yashgupta9330** [comment]
> I'm unable to run the frontend for this PR:
> 
> ```
> nest-frontend  | /home/owasp/node_modules/rollup/dist/native.js:63
> nest-frontend  |                throw new Error(
> nest-frontend  |                      ^
> nest-frontend  | 
> nest-frontend  | Error: Cannot find module @rollup/rollup-linux-arm64-gnu. npm has a bug related to optional dependencies (https://github.com/npm/cli/issues/4828). Please try `npm i` again after removing both package-lock.json and node_modules directory.
> nest-frontend  |     at requireWithFriendlyError (/home/owasp/node_modules/rollup/dist/native.js:63:9)
> nest-frontend  |     at Object.<anonymous> (/home/owasp/node_modules/rollup/dist/native.js:72:76)
> nest-frontend  |     at Module._compile (node:internal/modules/cjs/loader:1565:14)
> nest-frontend  |     at Object..js (node:internal/modules/cjs/loader:1708:10)
> nest-frontend  |     at Module.load (node:internal/modules/cjs/loader:1318:32)
> nest-frontend  |     at Function._load (node:internal/modules/cjs/loader:1128:12)
> nest-frontend  |     at TracingChannel.traceSync (node:diagnostics_channel:322:14)
> nest-frontend  |     at wrapModuleLoad (node:internal/modules/cjs/loader:219:24)
> nest-frontend  |     at cjsLoader (node:internal/modules/esm/translators:263:5)
> nest-frontend  |     at ModuleWrap.<anonymous> (node:internal/modules/esm/translators:196:7) {
> nest-frontend  |   [cause]: Error: Cannot find module '@rollup/rollup-linux-arm64-gnu'
> nest-frontend  |   Require stack:
> nest-frontend  |   - /home/owasp/node_modules/rollup/dist/native.js
> nest-frontend  |       at Function._resolveFilename (node:internal/modules/cjs/loader:1249:15)
> nest-frontend  |       at Function._load (node:internal/modules/cjs/loader:1075:27)
> nest-frontend  |       at TracingChannel.traceSync (node:diagnostics_channel:322:14)
> nest-frontend  |       at wrapModuleLoad (node:internal/modules/cjs/loader:219:24)
> nest-frontend  |       at Module.require (node:internal/modules/cjs/loader:1340:12)
> nest-frontend  |       at require (node:internal/modules/helpers:138:16)
> nest-frontend  |       at requireWithFriendlyError (/home/owasp/node_modules/rollup/dist/native.js:45:10)
> nest-frontend  |       at Object.<anonymous> (/home/owasp/node_modules/rollup/dist/native.js:72:76)
> nest-frontend  |       at Module._compile (node:internal/modules/cjs/loader:1565:14)
> nest-frontend  |       at Object..js (node:internal/modules/cjs/loader:1708:10) {
> nest-frontend  |     code: 'MODULE_NOT_FOUND',
> nest-frontend  |     requireStack: [ '/home/owasp/node_modules/rollup/dist/native.js' ]
> nest-frontend  |   }
> nest-frontend  | }
> nest-frontend  | 
> nest-frontend  | Node.js v22.12.0
> nest-frontend exited with code 1
> ```

Can you rerun it? I have undone the package-lock file

**Review by arkid15r — APPROVED** [review]
Thank you @Yashgupta9330 
Merging this now as we need this component for the new FE migration
